ext/drbbdup/drbbdup.c

```diff
@@ -0,0 +1,1198 @@
+/* **********************************************************
+ * Copyright (c) 2013-2020 Google, Inc.   All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "dr_defines.h"
+#include "dr_api.h"
+#include "drmgr.h"
+#include "drreg.h"
+#include "hashtable.h"
+#include "drbbdup.h"
+#include <string.h>
+#include <stdint.h>
+#include "../ext_utils.h"
+
+#ifdef DEBUG
+#    define ASSERT(x, msg) DR_ASSERT_MSG(x, msg)
+#    define LOG(dc, mask, level, ...) dr_log(dc, mask, level, __VA_ARGS__)
+#else
+#    define ASSERT(x, msg)
+#    define LOG(dc, mask, level, ...)
+#endif
+
+/**
+ * DynamoRIO Basic Block Duplication Extension: a code builder that
+ * duplicates code of basic blocks and dispatches control according to runtime conditions
+ * so that different instrumentation may be efficiently executed.
+ */
+
+#define HASH_BIT_TABLE 13
+
+/* THREAD SLOTS */
+#define DRBBDUP_ENCODING_SLOT 0
+#define DRBBDUP_XAX_REG_SLOT 1
+#define DRBBDUP_FLAG_REG_SLOT 2
+#define DRBBDUP_SLOT_COUNT \
+    DRBBDUP_FLAG_REG_SLOT + 1 /* Need to update if more slots are added. */
+
+/* A scratch register used by drbbdup's dispatcher. */
+#define DRBBDUP_SCRATCH_REG DR_REG_XAX
```

Comment:
So x86-only.  I would put in a #error under #ifndef X86 with a TODO i# comment up here.

Reply:
Where would this be placed exactly please? In the init function?

Reply:
Done?

Reply:
No #error would be at the top of the file (well, below the license header, etc.)

Reply:
Yea, that is how I did it.